public/app/features/alerting/unified/components/rule-editor/AlertRuleForm.tsx

```diff
@@ -9,98 +9,105 @@ import { DetailsStep } from './DetailsStep';
-  noDataState: GrafanaAlertState.NoData,
-  execErrState: GrafanaAlertState.Alerting,
-  evaluateEvery: '1m',
-  evaluateFor: '5m',
+import { RuleWithLocation } from 'app/types/unified-alerting';
+import { useDispatch } from 'react-redux';
+import { useCleanup } from 'app/core/hooks/useCleanup';
+import { rulerRuleToFormValues, defaultFormValues } from '../../utils/rule-form';
 
-  // system
-  expression: '',
-  forTime: 1,
-  forTimeUnit: 'm',
-});
+type Props = {
+  existing?: RuleWithLocation;
+};
 
-export const AlertRuleForm: FC<Props> = () => {
+export const AlertRuleForm: FC<Props> = ({ existing }) => {
   const styles = useStyles(getStyles);
   const dispatch = useDispatch();
 
-  useEffect(() => {
-    return () => {
-      dispatch(cleanUpAction({ stateSelector: (state) => state.unifiedAlerting.ruleForm }));
-    };
-  }, [dispatch]);
+  const defaultValues: RuleFormValues = useMemo(() => {
+    if (existing) {
+      return rulerRuleToFormValues(existing);
+    }
+    return defaultFormValues;
+  }, [existing]);
 
   const formAPI = useForm<RuleFormValues>({
     mode: 'onSubmit',
     defaultValues,
   });
 
-  const { handleSubmit, watch } = formAPI;
+  const { handleSubmit, watch, errors } = formAPI;
+
+  const hasErrors = !!Object.values(errors).filter((x) => !!x).length;
 
   const type = watch('type');
   const dataSourceName = watch('dataSourceName');
 
-  const showStep2 = Boolean(dataSourceName && type);
+  const showStep2 = Boolean(type && (type === RuleFormType.threshold || !!dataSourceName));
 
   const submitState = useUnifiedAlertingSelector((state) => state.ruleForm.saveRule) || initialAsyncRequestState;
+  useCleanup((state) => state.unifiedAlerting.ruleForm.saveRule);
 
-  const submit = (values: RuleFormValues) => {
+  const submit = (values: RuleFormValues, exitOnSave: boolean) => {
+    console.log('submit', values);
     dispatch(
       saveRuleFormAction({
-        ...values,
-        annotations: values.annotations.filter(({ key }) => !!key),
-        labels: values.labels.filter(({ key }) => !!key),
+        values: {
+          ...values,
+          annotations: values.annotations?.filter(({ key }) => !!key) ?? [],
+          labels: values.labels?.filter(({ key }) => !!key) ?? [],
+        },
+        existing,
+        exitOnSave,
       })
     );
   };
 
   return (
     <FormContext {...formAPI}>
-      <form onSubmit={handleSubmit(submit)} className={styles.form}>
+      <form onSubmit={handleSubmit((values) => submit(values, false))} className={styles.form}>
         <PageToolbar title="Create alert rule" pageIcon="bell" className={styles.toolbar}>
-          <ToolbarButton variant="default" disabled={submitState.loading}>
-            Cancel
-          </ToolbarButton>
-          <ToolbarButton variant="primary" type="submit" disabled={submitState.loading}>
+          <a href="/alerting/list">
+            <ToolbarButton variant="default" disabled={submitState.loading} type="button">
```

Comment:
Using an anchor will trigger a full page reload, if you use the react router navigation we'll stay inside the app.

```suggestion
            <ToolbarButton variant="default" disabled={submitState.loading} type="button" onClick={() => locationService.push('/alerting/list')}>
```

Reply:
Does not trigger page reload, there is a click handler on `body` element that intercepts anchror clicks. Could use  react router dom's `<Link` , but does not seem to be an accepted practice (yet)..

In general I think it's preferable to have an anchor to let browser, screen readers and etc know that it's really a link and will trigger navigation

Reply:
replaced anchor with `<Link />`, perhaps will look into creating `<LinkToolbarButton />` at some point